### PR TITLE
Improve Watcher test framework resiliency

### PR DESCRIPTION
**Pull request by AthenaEryma**
It is possible for the watches tracked by `ScheduleTriggerEngineMock` to
get out of sync with the Watches in the `ScheduleTriggerEngine`
production code, which can lead to watches failing to run.

This commit:
1. Changes `TimeWarp` to try to run the watch on all schedulers, rather than stopping after one which claims to have the watch registered. This reduces the impact of desynchronization between the mocking code and the backing production code.
2. Makes `ScheduleTriggerEngineMock` respect pauses of execution again. This is necessary to prevent duplicate watch invocations due to the above change. 
3. Tweaks how watches are registered in `ScheduleTriggerEngineMock` to prevent race conditions due to concurrent modification.
4. Tweaks `WatcherConcreteIndexTests` to use `TimeWarp` instead of waiting for watches to be triggered, as TimeWarp is more reliable and accomplishes the same goal.

This should fix:
https://github.com/elastic/elasticsearch/issues/35503
https://github.com/elastic/elasticsearch/issues/35506
https://github.com/elastic/elasticsearch/issues/40587
https://github.com/elastic/elasticsearch/issues/40631
https://github.com/elastic/elasticsearch/issues/40682 (this one is only muted on 6.7 so I'll have to test + unmute on the backport)

---

I've run the entire Watcher test suite with these changes ~1000 times and have only seen two failures, which were unrelated (e.g. `SocketTimeoutException`) , which is much more reliable than before these changes.

### CODE REVIEW

**Comment by gwbrown** [comment]
@elasticmachine run elasticsearch-ci/bwc
run elasticsearch-ci/2

**Comment by gwbrown** [comment]
@elasticmachine run elasticsearch-ci/2

Just running the tests a few times to make sure this is stable on CI.

**Comment by gwbrown** [comment]
I'm going to run these tests on CI several more times just to be sure, but I think this is ready for review and it's been very stable locally.

**Review by martijnvg — APPROVED** [review]
I think these are good improvements to the mock engine.

I think we should backport this change slowly, so that if unexpected failures occur in CI, it doesn't fail on all branches this change is targeted for.

**Comment by martijnvg** [comment]
@gwbrown I think this PR should also fix: #40631 Maybe unmute this test too in this pr?

**Comment by gwbrown** [comment]
Thanks for the review @martijnvg, and that's also a good idea to wait on the backports. I'll add the test you suggest and run the tests a couple more times before merging, then let it sit in master for a few days before backporting and check build-stats to make sure it's all good before backporting.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/test/AbstractWatcherIntegrationTestCase.java
Comment (jakelandis) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/test/AbstractWatcherIntegrationTestCase.java]: I don't think you need this `.filter`

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java
Comment (jakelandis) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java]: Curious why using an AtomicReference here ?

Comment (jakelandis) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java]: Nvmd.. I see now that you are swapping out the instances below. 

New question, why atomic swaps vs. shared lock (for this test code) ?

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java
Comment (jakelandis) [inline x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java]: this method by itself is not thread safe. I  would advise to synchronize this method, or use a shared lock between this start and stop, which negates the need for the atomic reference, or throw an exception if attempted to start/stop twice.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java
Comment (gwbrown) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java]: An earlier version of this used a lock and it made the tests very slow (this version takes about ~2-3m to run the whole suite, I killed the version with locks at ~10m), I'm not entirely sure why.

File: x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java
Comment (gwbrown) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java]: Good point, I'll take another look at this. There's a few cases where this could potentially have issues now that I look at it again.

Comment (martijnvg) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java]: This method is invoked from `TriggerService#start(...)` and that method is `synchronized`, so I don't think we need to `synchronize` this method.

Comment (gwbrown) [reply x-pack/plugin/watcher/src/test/java/org/elasticsearch/xpack/watcher/trigger/ScheduleTriggerEngineMock.java]: I've added `synchronized` to the methods which modify `watches` to ensure this is thread safe. How does this look @jakelandis?

**Review by jakelandis — APPROVED** [review]
LGTM

**Comment by gwbrown** [comment]
I'm seeing `watch was triggered on [0] schedulers, expected [1]` more often locally after the last round of changes and I'm not sure why, but I'm going to dig into that before merging this.

**Comment by gwbrown** [comment]
I've run the most recent revision (with a very minor tweak over the last reviewed version) over the weekend and it hasn't failed once in ~4000 runs, so I'm pretty confident in this version and intend to merge this after a couple successful CI runs.

**Comment by gwbrown** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by gwbrown** [comment]
@elasticmachine run elasticsearch-ci/2

Run was successful, just retriggering to get more confidence

**Comment by gwbrown** [comment]
These changes have been in master since the beginning of the week with no failures in the unmuted tests, so I've merged the backports as well.

**Pull request merged**
